src/Npgsql/SqlQueryParser.cs

```diff
@@ -148,17 +150,26 @@ internal void ParseRawQuery(string sql, bool standardConformantStrings, NpgsqlPa
                 if (currCharOfs >= end || !IsParamNameChar(ch = sql[currCharOfs])) {
                     var paramName = sql.Substring(currTokenBeg, currCharOfs - currTokenBeg);
 
-                    if (!_paramIndexMap.TryGetValue(paramName, out var index)) {
+                    if (!_paramIndexMap.TryGetValue(paramName, out var index))
+                    {
                         // Parameter hasn't been seen before in this query
                         if (!parameters.TryGetValue(paramName, out var parameter))
                         {
-                            _rewrittenSql.Append(paramName);
-                            currTokenBeg = currCharOfs;
-                            if (currCharOfs >= end)
-                                goto Finish;
-
-                            currCharOfs++;
-                            goto NoneContinue;
+                            if (deriveParameters)
```

Comment:
Are you sure?
From my understanding (and walking through the code with the debugger) the rewriting step follows a few lines below and only inserts the $1 ($2, $3, ...)= instead of the placeholder as it is required from the backend.

Reply:
Sorry, I misread the code...

The enclosing if is to check for cases where a placeholder (`@p`) is detected which does not have a corresponding NpgsqlParameter. For derivation this is the normal situation (unless a placeholder is specified twice), but for execution this is *usually* an error situation. At some point Npgsql actually raised an exception, but because the built-in SQL parser isn't perfect it used to identify certain valid SQL statements as containing parameters even though they did not. As a result, if a corresponding parameter instance isn't found, the placeholder (which may actually not be a placeholder at all) is left as-is.

These parser issues will become problematic when deriving, since the parser will always assume that the placeholders really are placeholders (no provided parameter instances to check). Since these cases are seemingly rare/exotic this shouldn't be a big problem.

We can leave this as-is.